### PR TITLE
Speaker Diarization Recipe for AMI Dataset

### PR DESCRIPTION
**Pull request by Qingzheng-Wang**
## What?

- Add a speaker diarization recipe for AMI dataset.
- Add the frame concatenation and subsampling in the `espnet2/diar/espnet_model.py`. 
- Add the missing `subsampling` argument to the command about `score_der.sh` in the scoring stage of `diar.sh` to ensure that the `subsampling` factor is correctly parsed by `score_der.sh`.

## Why?

- The AMI dataset is widely used for speaker diarization experiments, so supporting diarization on this dataset is crucial.
- The ESPNet diarization model currently lacks support for concatenating input frames with context frames and subsampling inputs, a technique used in the EEND-EDA paper.
- The `subsampling` argument is missing in the command about `score_der.sh` in the scoring stage of `diar.sh`. 

## See also

Issue #5993
PR #5969

### CODE REVIEW

File: egs2/TEMPLATE/diar1/diar.sh [hunk 1]
```diff
@@ -55,7 +55,7 @@ diar_config= # Config for diar model training.
 diar_args=   # Arguments for diar model training, e.g., "--max_epoch 10".
              # Note that it will overwrite args in diar config.
 feats_normalize=global_mvn # Normalizaton layer type.
-num_spk=2    # Number of speakers in the input audio
+num_spk=     # Number of speakers in the input audio
```
Comment (jctian98) [inline egs2/TEMPLATE/diar1/diar.sh:58]: is this change to the default value intended?

File: egs2/ami/diar1/assets/eend_der.png
Comment (jctian98) [inline egs2/ami/diar1/assets/eend_der.png:1]: We usually don't have these figures. Looks very good to include them!

File: egs2/TEMPLATE/diar1/diar.sh (hunk 1 above)
Comment (Qingzheng-Wang) [reply egs2/TEMPLATE/diar1/diar.sh:58]: I think it would be better to set the num_spk value in the run script since we cannot assume the dataset contains 2 speakers in default.

Comment (Qingzheng-Wang) [reply egs2/TEMPLATE/diar1/diar.sh:58]: Sorry, I just realized that this change may have an impact on other diar recipes which may not specify the num_spk in the run script (I checked librimix's diar run script). I will submit a new commit to remove this change.

File: egs2/ami/diar1/conf/train_diar.yaml [hunk 2]
```diff
@@ -0,0 +1,78 @@
+# The initial version is copy from egs2/librimix/diar1/conf/train_diar.yaml
+#    This config file is for SA-EEND.
+#    For the details about SA-EEND, refer to the following paper:
+#    SA-EEND: https://arxiv.org/pdf/1909.06247.pdf
+
```
Comment (jctian98) [inline egs2/ami/diar1/conf/train_diar.yaml:5]: It would be better to describe your computing usage, i.e., how many GPUs and how many hours.

File: egs2/ami/diar1/conf/train_diar_eda.yaml [hunk 3]
```diff
@@ -0,0 +1,85 @@
+#    The initial version of this file is copied from egs2/librimix/diar1/conf/train_diar_eda.yaml
+#    This config file is for EEND-EDA pre-training stage (training on fixed number of speakers).
+#    For the details about EEND-EDA, refer to the following papers:
+#    EEND-EDA: https://arxiv.org/pdf/2005.09921.pdf, https://arxiv.org/pdf/2106.10654.pdf
+
```
Comment (jctian98) [inline egs2/ami/diar1/conf/train_diar_eda.yaml:5]: Same here: your computing usage

File: egs2/ami/diar1/conf/train_diar_eda.yaml [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+# others:
+best_model_criterion:
+-  - valid
+   - acc
+   - max
+keep_nbest_models: 10
+
+# batch_type: numel
+# batch_bins: 15000000
+# accum_grad: 4
```
Comment (jctian98) [inline egs2/ami/diar1/conf/train_diar_eda.yaml:68]: can we clean up these commented lines? same for other config files.

File: egs2/ami/diar1/conf/train_diar_eda_adapt.yaml
Comment (jctian98) [inline egs2/ami/diar1/conf/train_diar_eda_adapt.yaml:1]: Sorry my knowledge for this task is limited. Could you explain a bit the different between this file and "train_diar_eda.yaml" ?

File: egs2/ami/diar1/conf/train_diar_eda.yaml (hunk 4 above)
Comment (Qingzheng-Wang) [reply egs2/ami/diar1/conf/train_diar_eda.yaml:68]: Sure!

File: egs2/ami/diar1/local/data.sh [hunk 5]
```diff
@@ -0,0 +1,164 @@
...
+log "data preparation started"
+
+# ================= Download AMI diarization setup files =================
+# AMI corpus for speaker diarization setup from gihub :
+# https://github.com/pyannote/AMI-diarization-setup/tree/main
+if [ ${stage} -le 0 ] && [ ${stop_stage} -ge 0 ] ; then
+    log "Start downloading AMI-diarization-setup from github"
+    # This is a fork created by Qingzheng Wang, mainly modified the database.yml,
+    # to adapt to ESPNet's directory setting
+    URL=https://github.com/Qingzheng-Wang/AMI-diarization-setup.git
```
Comment (jctian98) [inline egs2/ami/diar1/local/data.sh:76]: Please keep this repo public as other people may need it in the future :)

File: egs2/ami/diar1/conf/train_diar_eda_adapt.yaml
Comment (Qingzheng-Wang) [reply egs2/ami/diar1/conf/train_diar_eda_adapt.yaml:1]: Most of the utterances in AMI dataset are with 4 speakers, but several with 3 or 5 speakers. My training strategy is first train the model with 4 speaker data, and then adapt the model with 3 and 5 speaker data. train_diar_eda.yaml is for 4 speakers, train_diar_eda_adapt.yaml if for 3 or 5 speakers.

File: egs2/ami/diar1/local/prepare_kaldi_files.py [hunk 6]
```diff
@@ -0,0 +1,246 @@
+# Prepare Kaldi-style data directory for AMI diarization task
```
Comment (jctian98) [inline egs2/ami/diar1/local/prepare_kaldi_files.py:1]: need a header for the interpreter. You can also add your name to this file.

File: egs2/ami/diar1/local/data.sh (hunk 5 above)
Comment (Qingzheng-Wang) [reply egs2/ami/diar1/local/data.sh:76]: Sure!

**Comment by jctian98** [comment]
My knowledge of the diariazation task is limited, so I mainly checked the format. It's obvious that @Qingzheng-Wang has put a lot of effort into this PR, and it's of very high quality. 
Just one comment: since you revised the `espnet_model.py` file, could you also add the test cases for your modifications so that they can be covered by `codecov`? 
Many thanks!

File: egs2/ami/diar1/local/prepare_kaldi_files.py (hunk 6 above)
Comment (Qingzheng-Wang) [reply egs2/ami/diar1/local/prepare_kaldi_files.py:1]: Already updated here!

File: egs2/ami/diar1/conf/train_diar_eda.yaml (hunk 3 above)
Comment (Qingzheng-Wang) [reply egs2/ami/diar1/conf/train_diar_eda.yaml:6]: Already updated!

File: egs2/ami/diar1/conf/train_diar.yaml (hunk 2 above)
Comment (Qingzheng-Wang) [reply egs2/ami/diar1/conf/train_diar.yaml:6]: Already updated!

**Comment by Qingzheng-Wang** [comment]
> My knowledge of the diariazation task is limited, so I mainly checked the format. It's obvious that @Qingzheng-Wang has put a lot of effort into this PR, and it's of very high quality. Just one comment: since you revised the `espnet_model.py` file, could you also add the test cases for your modifications so that they can be covered by `codecov`? Many thanks!

Hi @jctian98, thank you for pointing that out! I have added the relevant test cases to `test/espnet2/diar/test_espnet_model.py` to cover the modifications made to `espnet_model.py`. All CI tests have now passed successfully.

**Comment by jctian98** [comment]
Thanks! @Qingzheng-Wang 
It looks good to me :)

**Comment by sw005320** [comment]
Thanks.
I think this PR is a very good status, but I have a concern that the DER is not very good.

@shota-horiguchi, do you think it is acceptable?
Note that this recipe only uses the AMI training data.

**Comment by shota-horiguchi** [comment]
> Thanks. I think this PR is a very good status, but I have a concern that the DER is not very good.
> 
> @shota-horiguchi, do you think it is acceptable? Note that this recipe only uses the AMI training data.

I have never trained a model sorely on AMI data, so I'm not sure about the validity of the DER.
But frankly, I think the reported DERs are not a very good value.
For reference, the DER for perfect voice activity detection (without speaker label, treating all speech segments are from the same speaker) is 60.59%.
I think it is possible that the reported DER could be achieved simply by training a voice activity detector.

**Comment by sw005320** [comment]
Yeah, I agree.
Do you think our network configuration is OK?
(It would be better to use 16 kHz and a larger filterbank size, but they would not change the result so much)
I'm also concerned about a way in this PR to handle the long recording.

@Qingzheng-Wang, can you revisit some setups?
https://github.com/espnet/espnet/blob/d9d3769bbe5897c1ef981c8dbc409aee4c771956/egs2/ami/diar1/README.md

**Comment by shota-horiguchi** [comment]
I have not checked in-depth, but the network config seems good.
The learning curve also appears reasonable, considering it wasn't pretrained with simulated data.

As Shinji mentioned, I suspect the issue lies in the inference process.
It's notable that there's a roughly 20% gap between the DERs on the learning curve and the final reported DERs.
For the AMI data, I used CPU inference since the recordings were too long to process on the GPU.

**Comment by Qingzheng-Wang** [comment]
Hi @shota-horiguchi and @sw005320, 

During both training and evaluation, I used a subsampling factor of 10. For evaluation, I configured the `espnet_model` to calculate the DER only on the subsampled frames. However, during inference, while I also applied the subsampling factor of 10, the scoring utilities at the inference stage assume that the neighboring frames of these subsampled frames share the same labels. This assumption led to the 20% gap between evaluation and inference.

**Comment by Qingzheng-Wang** [comment]
> Yeah, I agree. Do you think our network configuration is OK? (It would be better to use 16 kHz and a larger filterbank size, but they would not change the result so much) I'm also concerned about a way in this PR to handle the long recording.
> 
> @Qingzheng-Wang, can you revisit some setups? https://github.com/espnet/espnet/blob/d9d3769bbe5897c1ef981c8dbc409aee4c771956/egs2/ami/diar1/README.md

Hi professor, the inference stage is performed on the CPU using the entire test audio without segmentation or hopping. I've added this to the `README.md`
